File: pulumi/library/go/pkg/instance_template/instance_template_test.go
/*
 * Copyright 2026 Vitruvian Software
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package instance_template

import (
	"testing"

	"github.com/VitruvianSoftware/pulumi-library/go/internal/testutil"
	"github.com/pulumi/pulumi/sdk/v3/go/pulumi"
	"github.com/stretchr/testify/require"
)

func TestNewInstanceTemplate_NilArgs(t *testing.T) {
	err := pulumi.RunErr(func(ctx *pulumi.Context) error {
		_, err := NewInstanceTemplate(ctx, "test", nil)
		require.Error(t, err)
		require.Contains(t, err.Error(), "args is required")
		return nil
	}, pulumi.WithMocks("project", "stack", testutil.NewTracker()))
	require.NoError(t, err)
}

func TestNewInstanceTemplate_Basic(t *testing.T) {
	tracker := testutil.NewTracker()
	err := pulumi.RunErr(func(ctx *pulumi.Context) error {
		_, err := NewInstanceTemplate(ctx, "test-tmpl", &InstanceTemplateArgs{
			Project:     pulumi.String("test-proj"),
			Region:      "us-central1",
			SourceImage: "projects/debian-cloud/global/images/family/debian-11",
			Network:     pulumi.String("projects/test/global/networks/vpc"),
			Subnetwork:  pulumi.String("projects/test/regions/us-central1/subnetworks/subnet"),
		})
		require.NoError(t, err)
		return nil
	}, pulumi.WithMocks("project", "stack", tracker))
	require.NoError(t, err)

	tracker.RequireType(t, "gcp:compute/instanceTemplate:InstanceTemplate", 1)
}

func TestNewInstanceTemplate_WithShieldedVm(t *testing.T) {
	tracker := testutil.NewTracker()
	err := pulumi.RunErr(func(ctx *pulumi.Context) error {
		_, err := NewInstanceTemplate(ctx, "test-shielded", &InstanceTemplateArgs{
			Project:             pulumi.String("test-proj"),
			Region:              "us-central1",
			MachineType:         "n2-standard-4",
			SourceImage:         "projects/cos-cloud/global/images/family/cos-stable",
			DiskSizeGb:          50,
			DiskType:            "pd-ssd",
			Network:             pulumi.String("projects/test/global/networks/vpc"),
			Subnetwork:          pulumi.String("projects/test/regions/us-central1/subnetworks/subnet"),
			ServiceAccountEmail: pulumi.String("sa@test.iam.gserviceaccount.com"),
			Tags:                []string{"allow-ssh"},
			Labels:              map[string]string{"env": "prod"},
			Metadata:            map[string]string{"startup-script": "#!/bin/bash\necho hello"},
			EnableShieldedVm:    true,
		})
		require.NoError(t, err)
		return nil
	}, pulumi.WithMocks("project", "stack", tracker))
	require.NoError(t, err)

	tracker.RequireType(t, "gcp:compute/instanceTemplate:InstanceTemplate", 1)
}

func TestNewInstanceTemplate_ConfidentialVm(t *testing.T) {
	tracker := testutil.NewTracker()
	err := pulumi.RunErr(func(ctx *pulumi.Context) error {
		_, err := NewInstanceTemplate(ctx, "test-confidential", &InstanceTemplateArgs{
			Project:              pulumi.String("test-proj"),
			Region:               "us-central1",
			MachineType:          "n2d-standard-2",
			SourceImage:          "projects/ubuntu-os-cloud/global/images/family/ubuntu-2204-lts",
			Network:              pulumi.String("projects/test/global/networks/vpc"),
			Subnetwork:           pulumi.String("projects/test/regions/us-central1/subnetworks/subnet"),
			EnableConfidentialVm: true,
		})
		require.NoError(t, err)
		return nil
	}, pulumi.WithMocks("project", "stack", tracker))
	require.NoError(t, err)

	tracker.RequireType(t, "gcp:compute/instanceTemplate:InstanceTemplate", 1)
}

File: pulumi/library/ts/packages/instance-template/src/instance-template.test.ts
/**
 * Copyright 2026 Vitruvian Software
 *
 * Unit tests for the InstanceTemplate and ComputeInstance modules.
 */

import { describe, it, expect, beforeAll } from "vitest";
import { setupPulumiMocks } from "../../../test/mocks";

beforeAll(() => {
    setupPulumiMocks();
});

describe("InstanceTemplate", () => {
    it("should export the InstanceTemplate class", async () => {
        const mod = await import("./index");
        expect(mod.InstanceTemplate).toBeDefined();
    });

    it("should validate template args", async () => {
        const args = {
            projectId: "prj-test",
            region: "us-central1",
            machineType: "e2-medium",
            sourceImage: "debian-cloud/debian-11",
            network: "projects/prj-test/global/networks/vpc-test",
            subnetwork: "projects/prj-test/regions/us-central1/subnetworks/sb-test",
        };
        expect(args.machineType).toBe("e2-medium");
        expect(args.region).toBe("us-central1");
    });

    it("should instantiate InstanceTemplate", async () => {
        const mod = await import("./index");
        const tmpl = new mod.InstanceTemplate("test-template", {
            projectId: "prj-test",
            region: "us-central1",
            machineType: "f1-micro",
            sourceImage: "debian-cloud/debian-11",
            network: "vpc-test",
            subnetwork: "sb-test",
        });
        expect(tmpl).toBeDefined();
    });
});

describe("ComputeInstance", () => {
    it("should export the ComputeInstance class", async () => {
        const mod = await import("./index");
        expect(mod.ComputeInstance).toBeDefined();
    });

    it("should instantiate ComputeInstance", async () => {
        const mod = await import("./index");
        const instance = new mod.ComputeInstance("test-instance", {
            projectId: "prj-test",
            zone: "us-central1-a",
            instanceName: "test-vm",
            instanceTemplate: "projects/prj-test/global/instanceTemplates/tmpl-test",
            subnetwork: "projects/prj-test/regions/us-central1/subnetworks/sb-test",
        });
        expect(instance).toBeDefined();
    });
});
